tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php

```diff
@@ -1616,6 +1616,64 @@ public function testSchemaDiffForeignKeys(): void
             );
         }
     }
+
+    public function testUnnamedForeignKeyConstraintHandling(): void
+    {
+        $sm                   = $this->connection->getSchemaManager();
+        $unnamedFkReferenced1 = new Table('unnamedfk_referenced1');
+        $unnamedFkReferenced1->addColumn('id1', 'integer', ['not null' => true]);
+        $unnamedFkReferenced1->addColumn('id2', 'integer', ['not null' => true]);
+        $unnamedFkReferenced1->setPrimaryKey(['id1', 'id2']);
+
+        $unnamedFkReferenced2 = new Table('unnamedfk_referenced2');
+        $unnamedFkReferenced2->addColumn('id', 'integer', ['not null' => true]);
+        $unnamedFkReferenced2->setPrimaryKey(['id']);
+
+        $unnamedFkReferencing = new Table('unnamedfk_referencing');
+        $unnamedFkReferencing->addColumn('id', 'integer', ['not null' => true]);
+        $unnamedFkReferencing->setPrimaryKey(['id']);
+        $unnamedFkReferencing->addColumn('reference1_id1', 'integer', ['not null' => true]);
+        $unnamedFkReferencing->addColumn('reference1_id2', 'integer', ['not null' => true]);
+        $unnamedFkReferencing->addColumn('reference2_id', 'integer', ['not null' => true]);
+        $unnamedFkReferencing->addForeignKeyConstraint(
+            $unnamedFkReferenced1,
+            ['reference1_id1', 'reference1_id2'],
+            ['id1', 'id2'],
+            ['onDelete' => 'CASCADE']
+        );
+        $unnamedFkReferencing->addForeignKeyConstraint(
+            $unnamedFkReferenced2,
+            ['reference2_id'],
+            ['id'],
+            ['onDelete' => 'CASCADE']
+        );
+
+        $sm->dropAndCreateTable($unnamedFkReferenced1);
+        $sm->dropAndCreateTable($unnamedFkReferenced2);
+        $sm->dropAndCreateTable($unnamedFkReferencing);
+
+        $onlineTable = $sm->listTableDetails('unnamedfk_referencing');
+
+        $offlineTable = new Table('unnamedfk_referencing');
+        $offlineTable->addColumn('id', 'integer');
+        $offlineTable->addColumn('reference1_id1', 'integer');
+        $offlineTable->addColumn('reference1_id2', 'integer');
+        $offlineTable->addColumn('reference2_id', 'integer');
+
+        $comparator = new Comparator();
+        $diff       = $comparator->diffTable($offlineTable, $onlineTable);
```

Comment:
Not sure I understand the logic:

1. We compare the offline table (which doesn't have the foreign keys) with the online one (which does).
2. The diff will obviously have the missing foreign keys.
3. By executing the diff SQL, we apply this diff to the online schema (which already has the foreign keys) which fails.

What's the expected result?

Reply:
The only reason why the test passes on SQLite is that its schema manager, instead of adding FKs to the table, creates a new table and drops the existing one: https://github.com/doctrine/dbal/blob/b5e15999e8f626b9a50edef142fffc7be57d0211/lib/Doctrine/DBAL/Platforms/SqlitePlatform.php#L961-L981

Reply:
See https://github.com/doctrine/dbal/pull/4256/#discussion_r484050582 for the explanation about the logic . This is a resolved discussion, I'm not sure how Github behaves when you link to that, if it does not work it was my first comment on that PR.

>  it works anyway because we start by dropping the table

I see we came to the same conclusion. This logic is unlikely to work with other RDBMS than SQLite.

Reply:
I see that the updated #4256 no longer has this logic but if it did, it would mean we were fixing something that doesn't make sense in the first place (i.e. we didn't replicate the issue properly in the form of a test or it's not a BC break because, despite the fact it worked in an earlier version, it wasn't meant to).

Reply:
> This logic is unlikely to work with other RDBMS than SQLite since they support `ALTER TABLE` with foreign keys

Regardless of the underlying platform, I don't understand what this “logic” is about. If you compute a diff between `a` and `b` and then apply to `b`, what's the intent, and what is the expected result? Purely logically.

Reply:
Maybe @beberlei can answer that, to me it's not fully clear to be honest.